content/200-concepts/100-components/02-prisma-client/052-advanced-type-safety/index.mdx

```diff
@@ -227,6 +227,27 @@ prisma.post.create({
 
 </TabbedContent>
 
-This query will also result in an error if an author with an `id` of `1` does not exist. In this case, Prisma Client will give a more descriptive error message. You can also use the [`connectOrCreate`](../../../../reference/api-reference/prisma-client-reference#connectorcreate) API to safely create a new user if one does not already exist with the given `id`.
+This query will also result in an error if an author with an `id` of `1` does not exist. In this case, Prisma Client will give a more descriptive error message. You can also use the [`connectOrCreate`](/reference/api-reference/prisma-client-reference#connectorcreate) API to safely create a new user if one does not already exist with the given `id`.
 
 We recommend using the "safe" `Input` types whenever possible.
+
+## Type utilities
+
+<Admonition type = "info">
+
+This feature is available from Prisma version 4.9.0 upwards.
+
+</Admonition>
+
+To help you create highly type-safe applications, Prisma Client provides a set of type utilities that tap into input and output types. These types are fully dynamic, which means that they adapt to any given model and schema. You can use them to improve the auto-completion and developer experience of your projects.
+
+This is especially useful in [validating inputs](./prisma-validator) and [shared Prisma Client extensions](/concepts/components/prisma-client/client-extensions/shared-extensions).
+
+The following type utilities are available in Prisma Client:
+
+- `Exact<Input, Shape>`: Enforces strict type safety on `Input`. `Exact` makes sure that a generic type `Input` strictly complies with the type that you specify in `Shape`. It [narrows](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) `Input` down to the most precise types.
+- `Args<Type, Operation>`: Retrieves the input arguments for any given model and operation. This is particularly useful for extension authors who want to do the following:
+  - Re-use existing types to extend or modify them.
+  - Benefit from the same auto-completion experience as on existing operations.
+- `Result<Type, Arguments, Operation>`: Takes the input arguments and provides the result for a given model and operation. You would usually use this in conjunction with `Args`. As with `Args`, `Result` helps you to re-use existing types to extend or modify them.
+- `Payload<Type, Operation>`: Retrieves the entire structure of the result, as scalars and relations objects for a given model and operation. For example, you can use this to determine which keys are scalars or objects at a type level.
```

Comment:
Worth/possible to have usage examples of these?
I understand the words, but have no idea what to do with them.